crates/pkvdb/src/lib.rs

```diff
@@ -0,0 +1,451 @@
+    ) -> sp_std::result::Result<Option<sp_trie::DBValue>, sp_state_machine::DefaultError> {
+        Ok(hash_db::HashDB::get(self, key, prefix))
+    }
+}
+
+impl<H, T> Kvdb<H, T>
+where
+    H: Hasher,
+    T: Default + PartialEq<T> + AsRef<[u8]> + for<'a> From<&'a [u8]> + Clone + Send + Sync,
+{
+    pub fn with_null_node<P: AsRef<Path>>(path: P, null_key: &[u8], null_node_data: T) -> Self {
+        let mut options = LevelDBOptions::default();
+        options.env = Rc::new(Box::new(GramineEnv::new()));
+        let db = DB::open(path, options).unwrap();
+        Kvdb {
+            leveldb: Arc::new(Mutex::new(db)),
+            hashed_null_node: H::hash(null_key),
+            null_node_data,
+        }
+    }
+
+    pub fn new<P: AsRef<Path>>(path: P) -> Self {
+        Self::with_null_node(path, &[0u8][..], [0u8][..].into())
+    }
+
+    pub fn purge(&mut self) {
+        // TODO:george this approach is not sound for performance
+        if let Ok(mut db_mut) = self.leveldb.lock() {
+            let mut writebatch = WriteBatch::new();
+            if let Ok(mut iter) = DB::new_iter(db_mut.borrow_mut()) {
+                while let Some((key, val)) = iter.next() {
+                    // FIXME: if we should keep the negative reference count
+                    if i32::from_be_bytes(val[0..4].try_into().unwrap()) == 0 {
+                        writebatch.delete(&key);
+                    }
+                }
+                let _ = DB::write(db_mut.borrow_mut(), writebatch, true);
+            }
+        }
+    }
+
+    // only for test
+    pub fn clear(&mut self) {
+        if let Ok(mut db_mut) = self.leveldb.lock() {
+            let mut writebatch = WriteBatch::new();
+            if let Ok(mut iter) = DB::new_iter(db_mut.borrow_mut()) {
+                while let Some((key, _)) = iter.next() {
+                    writebatch.delete(&key);
+                }
+                let _ = DB::write(db_mut.borrow_mut(), writebatch, true);
+            }
+        }
+    }
+
+    pub fn raw(&self, key: &H::Out, _prefix: Prefix) -> Option<(T, i32)> {
+        if key == &self.hashed_null_node {
+            return Some((self.null_node_data.clone(), 1));
+        }
+        let mut db_mut = self.leveldb.lock().unwrap();
+        match DB::get(db_mut.borrow_mut(), key.as_ref()).map(Inner::<T>::from) {
+            Some(Inner(value, rc)) => Some((value, rc)),
+            _ => None,
+        }
+    }
+}
+
+impl<H> Kvdb<H, sp_trie::DBValue>
+where
+    H: Hasher,
+{
+    pub fn consolidate(&mut self, mut transaction: Transaction<H>) {
+        if let Ok(mut db_mut) = self.leveldb.lock() {
+            let mut writebatch = WriteBatch::new();
+
+            for (key, (value, rc)) in transaction.drain() {
+                match DB::get(db_mut.borrow_mut(), key.as_ref())
+                    .map(Inner::<sp_trie::DBValue>::from)
+                {
+                    Some(mut inner) => {
+                        inner.1 += rc;
```

Comment:
When to decrease the refcount?

Reply:
rc variable could be a negative number then will reduce the inner.rc